lib/ansible/modules/windows/win_dns_client.ps1

```diff
@@ -108,40 +113,55 @@ If(-not $(Get-Command Set-DnsClientServerAddress -ErrorAction SilentlyContinue))
     New-Alias Set-DnsClientServerAddress Set-DnsClientServerAddressLegacy
 }
 
-Function Get-DnsClientMatch {
+Function Test-DnsClientMatch {
     Param(
         [string] $adapter_name,
-        [string[]] $ipv4_addresses
+        [string[]] $dns_servers
     )
+    $dns_servers = $dns_servers | Where-Object {
+        (Assert-IPAddress $_) -and ([System.Net.IPAddress]$_ in $AddressFamilies)
+    } | ForEach-Object {
+        [System.Net.IPAddress]$_
+    }
 
     Write-DebugLog ("Getting DNS config for adapter {0}" -f $adapter_name)
+    Write-DebugLog ("Current DNS settings: " + $($current_dns_all.IPAddressToString | Out-String))
 
-    $current_dns_all = Get-DnsClientServerAddress -InterfaceAlias $adapter_name
-
-    Write-DebugLog ("Current DNS settings: " + $($current_dns_all | Out-String))
-
-    $current_dns_v4 = ($current_dns_all | Where-Object AddressFamily -eq 2 <# IPv4 #>).ServerAddresses
-
-    If (($null -eq $current_dns_v4) -and ($null -eq $ipv4_addresses)) {
-        $v4_match = $True
-    }
+    $current_dns = [System.Net.IPAddress[]](
+        Get-DnsClientServerAddress -InterfaceAlias $adapter_name | ForEach-Object {
+            $_.ServerAddresses
+        } | Where-Object {
+            (Assert-IPAddress $_) -and ($_.AddressFamily -in $AddressFamilies)
+        }
+    )
 
-    ElseIf (($null -eq $current_dns_v4) -or ($null -eq $ipv4_addresses)) {
-        $v4_match = $False
+    If (($null -eq $current_dns) -and ($null -eq $dns_servers)) {
+        Write-DebugLog "Neither are dns servers configured nor specified within the playbook."
+        return $true
+    } elseif ($null -eq $current_dns) {
+        Write-DebugLog "There are currently no dns servers specified, but they should be present."
+        return $false
+    } elseif ($null -eq $dns_servers) {
+        Write-DebugLog "There are currently dns servers specified, but they should be absent."
+        return $false
+    } else {}
+    foreach($address in $current_dns) {   
```

Comment:
Consider `Compare-Object` instead of all these loops

Reply:
How would I write this foreach with `compare-object`?

Reply:
I honestly don't remember what I meant